py3status/modules/speedtest.py

```diff
@@ -0,0 +1,321 @@
+                self.command += " --no-upload"
+
+            if not any(key.startswith("download") for key in self.placeholders):
+                self.command += " --no-download"
+
+        if self.server_id and int(self.server_id):
+            self.command += " --server " + str(self.server_id)
+
+        if self.button_share and all(
+            x in self.placeholders for x in ["download", "upload"]
+        ):
+            self.command += " --share"
+
+        self.command += " --timeout " + str(self.timeout)
+
+    def _is_running(self):
+        try:
+            self.py3.command_output(["pgrep", "speedtest-cli"])
+            return True
+        except self.py3.CommandError:
+            return False
+
+    def _cast_number(self, data):
+        try:
+            data = int(data)
+        except ValueError:
+            try:
+                data = float(data)
+            except ValueError:
+                pass
+        return data
+
+    def _get_speedtest_data(self):
+        return loads(self.py3.command_output(self.command)) or None
+
+    def _start(self):
+        self.cached_until = self.py3.time_in(self.sleep_timeout)
+
+        if not self._is_running():
+            self.cached_until = self.py3.CACHE_FOREVER
+            previous_data = self.py3.storage_get("speedtest_data") or {}
+            current_data = self._get_speedtest_data() or {}
+
+            if len(current_data) > 1:
+                # create a "total" for know if cnx is better or not
+                # between two run
+                current_data["total"] = int(current_data.get("download", 0)) + int(
+                    current_data.get("upload", 0)
+                )
+
+                # create "quality" #maybe bad name
+                if "total" in previous_data:
+                    if current_data["total"] >= previous_data["total"]:
+                        quality_key = 4
+                    else:
+                        quality_key = 3
+                else:
+                    if current_data["total"] <= 0:
+                        quality_key = 1
+                    else:
+                        quality_key = 2
+                current_data["quality"] = quality_key
+                current_data["quality_name"] = self.quality[quality_key]
+
+        # zero-ing if not fetched, raw version and units convertion
+        for x in ["download", "upload", "bytes_received", "bytes_sent"]:
+            unit = self.unit_size if "bytes" in x else self.unit_bitrate
+            current_data[x] = current_data.get(x, 0)
+            current_data[x + "_raw"] = current_data[x]
+            current_data[x], current_data[x + "_unit"] = self.py3.format_units(
+                current_data[x], unit=unit, si=self.si_units
+            )
+
+        # extra data, not sure we want to expose
+        current_data.update(self.py3.flatten_dict(current_data, delimiter='_'))
+        for x in ['client', 'server']:
+            del current_data[x]
```

Comment:
I think you might be creating `flatten_dict(current_data, delimiter='_', intermediates=True)`, then you delete the unwanted dicts. Does `current_data = self.py3.flatten...` works the same here?

Reply:
not need intermediates ;)